// @tldraw/lfg/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.ts
var src_exports = {};
__export(src_exports, {
  lfg: () => lfg
});
module.exports = __toCommonJS(src_exports);

// src/lfg.ts
var import_fs2 = __toESM(require("fs"));
var import_path3 = __toESM(require("path"));

// src/buildLibrary.ts
var import_fs = __toESM(require("fs"));
var import_path = __toESM(require("path"));
var import_esbuild = require("esbuild");
var import_zlib = require("zlib");
var import_child_process = require("child_process");
var import_dotenv = __toESM(require("dotenv"));
import_dotenv.default.config();
var { log } = console;
function buildLibrary({
  name,
  isNode,
  outdir,
  tsconfig,
  external,
  format,
  target,
  define,
  entryPoints,
  calculateSize,
  customEsbuildConfig,
  buildCss
}) {
  return new Promise((resolve) => {
    log(`\u2023 ${name}: Building packages.`);
    try {
      let finish = function(types = false, build3 = false) {
        if (build3)
          didBuild = true;
        if (types)
          didTypes = true;
        if (didBuild && didTypes) {
          log(`\u2714 ${name}: Build complete.`);
          resolve();
        }
      };
      let didBuild = false;
      let didTypes = false;
      const ts = (0, import_child_process.spawn)(`tsc`, [`--project`, tsconfig, `--outDir`, outdir], { shell: true });
      ts.stdout.on("data", function(data) {
        const str = data.toString();
        log(`\u2022 ${name}: ${str}`);
      });
      ts.on("exit", () => {
        var _a, _b;
        log(`\u2022 ${name}: Built types.`);
        const base_cfg = require(import_path.default.join(process.cwd(), "tsconfig.json"));
        let cfgWithPaths = "";
        if ((_a = base_cfg == null ? void 0 : base_cfg.compilerOptions) == null ? void 0 : _a.paths) {
          cfgWithPaths = "tsconfig.json";
        } else {
          const cfg = require(tsconfig);
          if ((_b = cfg == null ? void 0 : cfg.compilerOptions) == null ? void 0 : _b.paths) {
            cfgWithPaths = tsconfig;
          }
        }
        if (cfgWithPaths) {
          const trp = (0, import_child_process.spawn)(`tsconfig-replace-paths`, [`-p`, cfgWithPaths], { shell: true });
          trp.on("exit", () => {
            log(`\u2022 ${name}: Resolved paths.`);
            finish(true);
          });
        } else {
          finish(true);
        }
      });
      Promise.all((Array.isArray(format) ? format : [format]).map((fmt) => __async(this, null, function* () {
        const extension = fmt === "esm" ? ".mjs" : ".js";
        const buildResult = yield (0, import_esbuild.build)(__spreadValues({
          entryPoints,
          outdir,
          tsconfig,
          external,
          define,
          format: fmt,
          platform: isNode ? "node" : "neutral",
          outExtension: { ".js": extension },
          bundle: true,
          metafile: true,
          minify: true,
          sourcemap: true,
          treeShaking: true,
          target
        }, customEsbuildConfig));
        if (calculateSize) {
          import_fs.default.readFile(import_path.default.join(outdir, `index${extension}`), (_err, data) => {
            (0, import_zlib.gzip)(data, (_err2, result) => {
              const size = Object.values(buildResult.metafile.outputs).reduce((acc, { bytes }) => acc + bytes, 0);
              log(`\u2022 ${name}: Built ${fmt} package. (${(size / 1e3).toFixed(1)}kb / ${(result.length / 1e3).toFixed(1)}kb gzipped)`);
            });
          });
        }
      }))).then(() => {
        if (buildCss) {
          (0, import_esbuild.build)({
            entryPoints: buildCss.entryPoints,
            outfile: import_path.default.join(outdir, "index.css"),
            bundle: true,
            minify: true
          }).then(() => {
            finish(false, true);
          });
        } else {
          finish(false, true);
        }
      });
    } catch (e) {
      log(`x ${name}: Build failed due to an error. ${e.message}`);
    }
  });
}

// src/devLibrary.ts
var import_esbuild2 = require("esbuild");
var import_child_process2 = require("child_process");
var import_path2 = __toESM(require("path"));
var { log: log2 } = console;
var ERROR_REGEX = /merror/gm;
var CLEAN_REGEX = /Found 0 errors/gm;
var CHANGE_REGEX = /Starting incremental compilation/gm;
function devLibrary(_0) {
  return __async(this, arguments, function* ({
    name,
    isNode,
    outdir,
    tsconfig,
    external,
    target,
    define,
    format,
    entryPoints,
    customEsbuildConfig,
    buildCss
  }) {
    log2(`\u2023 ${name}: Starting watch mode.`);
    const ts = (0, import_child_process2.spawn)(`tsc`, [`-w`, `--project`, tsconfig, `--pretty`, `--preserveWatchOutput`], {
      shell: true
    });
    let state = "clean";
    ts.stdout.on("data", function(data) {
      const str = data.toString();
      if (CHANGE_REGEX.test(str))
        return;
      const isErrorMessage = ERROR_REGEX.test(str);
      const isCleanMessage = CLEAN_REGEX.test(str);
      if (isErrorMessage) {
        state = "error";
        log2(`\u2023 ${name}: Rebuilt types. Encountered errors:`);
      }
      switch (state) {
        case "error": {
          if (isCleanMessage) {
            state = "clean";
            log2(`\u2023 ${name}: Rebuilt types. TypeScript errors fixed.`);
          } else {
            str.split("\n").filter(Boolean).forEach((s) => log2(`\u2023 ${name}: ${s}`));
          }
          break;
        }
        case "clean": {
          if (isCleanMessage) {
            log2(`\u2023 ${name}: Rebuilt types.`);
          } else {
          }
          break;
        }
      }
    });
    try {
      log2(`\u2023 ${name}: Starting incremental build.`);
      const formats = Array.isArray(format) ? format : [format];
      formats.forEach((fmt) => {
        const extension = fmt === "esm" ? ".mjs" : ".js";
        (0, import_esbuild2.build)(__spreadProps(__spreadValues({
          entryPoints,
          outdir,
          tsconfig,
          external,
          define,
          format: fmt,
          target,
          platform: isNode ? "node" : "neutral",
          outExtension: { ".js": extension },
          minify: false,
          bundle: true,
          treeShaking: true,
          metafile: true,
          sourcemap: true,
          incremental: true
        }, customEsbuildConfig), {
          watch: {
            onRebuild(err) {
              if (err) {
                throw Error(`\u2023 ${name}: Rebuild failed.`);
              }
              log2(`\u2714 ${name}: Rebuilt package (${fmt}).`);
            }
          }
        }));
      });
      if (buildCss) {
        (0, import_esbuild2.build)({
          entryPoints: buildCss.entryPoints,
          outfile: import_path2.default.join(outdir, "index.css"),
          bundle: true,
          minify: false,
          incremental: true,
          watch: {
            onRebuild(err) {
              if (err) {
                throw Error(`\u2023 ${name}: Failed to rebuild CSS.`);
              }
              log2(`\u2714 ${name}: Rebuilt CSS.`);
            }
          }
        });
      }
    } catch (err) {
      ts == null ? void 0 : ts.kill();
      process.exit(1);
    }
  });
}

// src/lfg.ts
function lfg() {
  return __async(this, arguments, function* (opts = {}) {
    var _a, _b, _c;
    const {
      isDev = false,
      isNode = false,
      entryPoints = ["src/index.ts"],
      clean = true,
      devConfig = "tsconfig.dev.json",
      buildConfig = "tsconfig.build.json",
      devFormat = "esm",
      format = ["cjs", "esm"],
      target = "es6",
      outDir = "dist",
      external = {
        dependencies: true,
        devDependencies: true,
        peerDependencies: true
      },
      define = {
        "process.env.NODE_ENV": isDev ? '"development"' : '"production"'
      },
      calculateSize = true,
      customEsbuildConfig = {},
      buildCss
    } = opts;
    const cwd = process.cwd();
    const pkg = require(import_path3.default.join(cwd, "package.json"));
    const externals = [
      ...external.dependencies ? Object.keys((_a = pkg.dependencies) != null ? _a : {}) : [],
      ...external.devDependencies ? Object.keys((_b = pkg.devDependencies) != null ? _b : {}) : [],
      ...external.peerDependencies ? Object.keys((_c = pkg.peerDependencies) != null ? _c : {}) : []
    ];
    const outDirAbs = import_path3.default.join(cwd, outDir);
    let configAbs = import_path3.default.join(cwd, isDev ? devConfig : buildConfig);
    if (!import_fs2.default.existsSync(configAbs)) {
      const { log: log3 } = console;
      log3(`\u2023 ${pkg.name}: Could not find ${isDev ? "dev" : "build"} config file (${isDev ? devConfig : buildConfig}), using tsconfig.json.`);
      configAbs = import_path3.default.join(cwd, "tsconfig.json");
    }
    const entryPointsAbs = entryPoints.map((entryPoint) => import_path3.default.join(cwd, entryPoint));
    if (isDev) {
      yield buildLibrary({
        name: pkg.name,
        outdir: outDirAbs,
        tsconfig: configAbs,
        external: externals,
        entryPoints: entryPointsAbs,
        format,
        target,
        isNode,
        define,
        calculateSize,
        customEsbuildConfig,
        buildCss
      });
      return devLibrary({
        name: pkg.name,
        outdir: outDirAbs,
        tsconfig: configAbs,
        external: externals,
        format: devFormat,
        target,
        entryPoints: entryPointsAbs,
        isNode,
        define,
        calculateSize,
        customEsbuildConfig,
        buildCss
      });
    } else {
      if (clean) {
        if (import_fs2.default.existsSync(outDirAbs)) {
          import_fs2.default.rmSync(outDirAbs, { recursive: true });
        }
      }
      return buildLibrary({
        name: pkg.name,
        outdir: outDirAbs,
        tsconfig: configAbs,
        external: externals,
        entryPoints: entryPointsAbs,
        format,
        target,
        isNode,
        define,
        calculateSize,
        customEsbuildConfig,
        buildCss
      });
    }
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  lfg
});
